src/gui/servernotificationhandler.cpp

```diff
@@ -70,42 +70,41 @@ void ServerNotificationHandler::slotNotificationsReceived(const QJsonDocument &j
-        a._type = Activity::NotificationType;
-        a._accName = ai->account()->displayName();
-        a._id = json.value("notification_id").toInt();
-        a._subject = json.value("subject").toString();
-        a._message = json.value("message").toString();
-        QString s = json.value("link").toString();
-        if (!s.isEmpty()) {
-            a._link = QUrl(s);
-        }
-        a._dateTime = QDateTime::fromString(json.value("datetime").toString(), Qt::ISODate);
-
-        auto actions = json.value("actions").toArray();
-        foreach (auto action, actions) {
-            auto actionJson = action.toObject();
+    list.reserve(notifies.size());
+    for (const auto &element : notifies) {
+        const auto json = element.toObject();
+        Activity a(
+            Activity::NotificationType,
+            json.value(QStringLiteral("notification_id")).toVariant().value<Activity::Identifier>(),
+            ai->account(),
+            json.value(QStringLiteral("subject")).toString(),
+            json.value(QStringLiteral("message")).toString(),
+            QString(),
+            QUrl(json.value(QStringLiteral("link")).toString()),
+            QDateTime::fromString(json.value(QStringLiteral("datetime")).toString(), Qt::ISODate));
+
+        const auto actions = json.value("actions").toArray();
+        QVector<ActivityLink> linkList;
+        linkList.reserve(actions.size() + 1);
+        for (const auto &action : actions) {
+            const auto actionJson = action.toObject();
             ActivityLink al;
-            al._label = QUrl::fromPercentEncoding(actionJson.value("label").toString().toUtf8());
-            al._link = actionJson.value("link").toString();
-            al._verb = actionJson.value("type").toString().toUtf8();
-            al._isPrimary = actionJson.value("primary").toBool();
-
-            a._links.append(al);
+            al._label = QUrl::fromPercentEncoding(actionJson.value(QStringLiteral("label")).toString().toUtf8());
+            al._link = actionJson.value(QStringLiteral("link")).toString();
+            al._verb = actionJson.value(QStringLiteral("type")).toString().toUtf8();
+            al._isPrimary = actionJson.value(QStringLiteral("primary")).toBool();
+            linkList.append(al);
         }
 
         // Add another action to dismiss notification on server
         // https://github.com/owncloud/notifications/blob/master/docs/ocs-endpoint-v1.md#deleting-a-notification-for-a-user
         ActivityLink al;
         al._label = tr("Dismiss");
-        al._link  = Utility::concatUrlPath(ai->account()->url(), notificationsPath + "/" + QString::number(a._id)).toString();
+        al._link = Utility::concatUrlPath(ai->account()->url(), notificationsPath + "/" + QString::number(a.id())).toString();
         al._verb  = "DELETE";
         al._isPrimary = false;
-        a._links.append(al);
-
+        linkList.append(al);
+        a.setLinks(linkList);
```

Comment:
How about adding a method that returns a mutable reference on the list?

Reply:
hmm well I added the functions to make it more clear who modifies what.
A non const ref would break that, I don't like to have a setter here but the alternative would be a separate constructor...

Reply:
Makes sense. I'd go for the second constructor then, I guess. It's more laborious but also cleaner in the end, as your object then implements the value object pattern.

Reply:
Perhaps the other constructor is the way to go then? The current constructor offers the option to pass a single link, but I think it's perfectly viable to just accept a list, then change all existing code to pass a single-element list instead.

This way, you can implement the value object pattern, and make the objects basically immutable, which might come in handy in the future.

Reply:
If you want to stick with a setter, you should probably accept r-value references, and use `std::move` to pass the list. This avoids unnecessary copying.